identity-core/src/crypto/key/pair.rs

```diff
@@ -12,9 +13,11 @@ use crate::crypto::PrivateKey;
 use crate::crypto::PublicKey;
 use crate::error::Result;
 use crate::utils::generate_ed25519_keypair;
-use crate::utils::keypair_from_ed25519_private_key;
 
 /// A convenient type for representing a pair of cryptographic keys.
+// TODO: refactor with exact types for each key type? E.g. Ed25519KeyPair, X25519KeyPair etc.
+//       Maybe a KeyPair trait with associated types? Might need typed key structs
+//       like Ed25519Public, X25519Private etc. to avoid exposing pre-1.0 dependency types.
```

Comment:
Yeah I'm hesitant to do a `KeyPair` refactor at this point since it has sweeping implications everywhere. One concern I have is that with introducing specific types we multiply the number of structs we have to make bindings for.

E.g. now we have `KeyPair` with `Public` and `Private` keys represented as strings/`UInt8Array` soon, so basically one type exposed to the bindings. With a refactor we might have 6 different structs:

- `Ed25519KeyPair`
- `Ed25519Private`
- `Ed25519Public`
- `X25519KeyPair`
- `X25519Private`
- `X25519Public`

And more for each new `KeyType` we choose to support. Note that the specific types for keys may be important for places where we only want/need the public key, such as creating a new verification method. We've largely moved away from requiring `KeyPairs` in function parameters where the private key is not required.

If we do go ahead with such a refactor I think a `KeyPair` trait would be useful for generalization, e.g.

```rust
trait KeyPair {
  type Public;
  type Private;

  fn public(&self) -> &Self::Public;
  fn private(&self) -> &Self::Private;
}
```

I don't think that will be in this PR, however.

Reply:
Thanks for the explanations. It's fine to leave this for a future PR. Turning `KeyPair` into a trait is what I had in mind, too, that sounds good! I agree that we need to carefully consider the impact before going ahead, specifically for Wasm.